pkg/frontend/querymiddleware/model_extra.go

```diff
@@ -42,6 +42,24 @@ func newEmptyPrometheusResponse() *PrometheusResponse {
 	}
 }
 
+// GetMinT returns the minimum timestamp in milliseconds of data to be queried,
+// as determined from the start timestamp and any range vector or offset in the query.
+func (q *PrometheusRangeQueryRequest) GetMinT() (int64, error) {
+	minT, _, err := decodeQueryMinMaxTime(
```

Comment:
The main design decision I struggled with is whether to re-parse and re-calculate the query to determine MinT and MaxT every time they are called, or to set them only when the Start, End, and Query are set, since MinT and MaxT are derived form Start, End, and Query.

The deciding factor was that because these codec concrete types are implemented as protos, the fields are all public and therefore we cannot stop a consumer from setting MinT and MaxT incorrectly and violating the relationship.

I also assumed that it would take a lot of re-parsing to create a bottleneck anywhere close to the bottleneck of actual data fetching and query processing.

Reply:
Could we only parse the query once? As it stands, we'll parse it for each call to `GetMinT()` and `GetMaxT()`. Perhaps we can track whether we've computed `minT` and `maxT` on `PrometheusRangeQueryRequest` somehow, and only compute them if it hasn't been done already?

(same feedback applies to `PrometheusInstantQueryRequest`)